_posts/2016-05-12-JS-Conf-Budapest-day-1.md

```diff
@@ -0,0 +1,775 @@
+Let's dive in.
+
+1. First npm will take a look at your local cache and see if the package your are looking for is present.
+2. Next it will resort to the CDN network and will use a server whos is the closest to your position as possible.
+3. Finally, if npm can't find the package in local cache or the CDN network, it will look up in the registry. The registry is a set of servers all around the world and will try to match the best version that you are looking for.
+
+#### EACCESS error
+A lot of people have issues with EACCESS errors because they used sudo to install things.
+The easy solution is to always keep on using sudo BUT we can [easily fix npm permission issues](https://docs.npmjs.com/getting-started/fixing-npm-permissions).
+
+#### package.json
+Don't write your `package.json` yourself, let NPM do it!
+It will always do it better. Use `npm init` which will ask some basic questions and will generate `package.json` for you.
+
+#### Scopes
+A new feature in npm is scopes.
+These are modules that are "scoped" under an organization name that begins with `@`.
+Scopes can be public and private.
+Here is how to use scopes:
+
+{% highlight sh %}
+npm init --scope=username
+npm install @myusername/mypackage
+require('@myusername/mypackage')
+{% endhighlight %}
+
+#### npm-init.js
+To extend the `npm init` command it is possible to create an `npm-init.js` file.
+This file is a module that will be loaded by the `npm init` command and look in this file and provide basic configurations for the setup.
+By default the file is placed in root of your project: `~/.npm-init.js`.
+You can use [PromZard](https://github.com/npm/promzard) to ask questions to the user and perform logic based on the answers.
+Remember that `npm init` can always be re-run.
+
+#### Why add stuff in devDependencies.
+Simple: because production will install faster! A lot of people don't tend to do this, so please do this!
+When using this you can simple run the command below on production and be done with it.
+
+{% highlight sh %}
+npm install --production
+{% endhighlight %}
+
+#### Bundled dependencies
+One of the biggest problems right now with Node.js is how fast it is changing.
+This means that production systems can be very fragile and an `npm update` can easily break things.
+Using `bundledDependencies` is a way to get round this issue by ensuring, that you will always deliver the correct dependencies no matter what else may be changing.
+You can also use this to bundle up your own, private bundles and deliver them with the install.
+
+{% highlight sh %}
+npm install --save --save-bundle
+{% endhighlight %}
+
+#### Offline installs
+A way to prevent npm to look up the registry and ensure local installs is by adding the variable `--cache-min` and to set it to a high value such as 999999.
+
+{% highlight sh %}
+npm install --cache-min 999999
+{% endhighlight %}
+
+#### Run scripts
+In `package.json` it is possible to define default run scripts as showed below.
```

Comment:
In `package.json`**,(comma)** it is possible to define default run scripts as ~~showed~~ **shown** below.